Add NEON acceleration for `Vectorized<int[8|16|32|64>`

Nicoshev opened the pull request:
Summary:
Adding NEON specializations of Vectorized<T> for int8, int16, int32 and int64.

Correcness has been checked using test_ops.py and the comprehensive torch test

operator_benchmark_test.py has been enhanced by adding cases of bitwise operations, boolean ops and integer ops.
The benchmark, which uses the PyTorch API, shows significant enhancements in a wide variety of operations:

Before:

bitwise xor: 779.882us
boolean any: 636.209us
boolean all: 538.621us
integer mul: 304.457us
integer asr: 447.997us

After:

bitwise xor: 680.221us ---> 15% higher throughput
boolean any: 391.468us ---> 63% higher throughput
boolean all: 390.189us ---> 38% higher throughput
integer mul: 193.532us ---> 57% higher throughput
integer asr: 179.929us---> 149% higher throughput

Test Plan:
Correctness:

buck2 test @mode/opt //caffe2/test:test_ops
buck2 test @mode/opt //caffe2/test:torch
buck2 test @mode/opt //caffe2/test/distributed/launcher/fb:fb_run_test

Performance:

buck2 run mode/opt //caffe2/benchmarks/operator_benchmark/fb:operator_benchmark_test

Differential Revision: D84424638




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @snadampal @milpuz01 @nikhil-arm @fadara01

malfet commented on benchmarks/operator_benchmark/pt/boolean_test.py:
```diff
@@ -0,0 +1,73 @@
+
+
+all_short_configs = op_bench.config_list(
+    attr_names=["M", "N", "K"],
```
Considering that this benchmark are of reduction/elementwise ops, why do you care about 3 dims there

malfet commented:
@pytorchbot merge -f "Will update operator benchmarks async"

Nicoshev committed:
[Pytorch] Add NEON Vectorized<int> family of translation layers (#165273)

Summary:

Adding NEON specializations of Vectorized<T> for int8, int16, int32 and int64.
Some operations have been enhanced with SVE, whenever compilation options allow it.

Correcness has been checked using test_ops.py

operator_benchmark_test.py has been enhanced by adding cases of bitwise operations, boolean ops and integer ops.
The benchmark, which uses the PyTorch API, shows significant enhancements in a wide variety of operations:

Before:

bitwise xor: 779.882us
boolean any: 636.209us
boolean all: 538.621us
integer mul: 304.457us
integer lsr: 447.997us

After:

bitwise xor: 680.221us ---> 15% higher throughput
boolean any: 391.468us ---> 63% higher throughput
boolean all: 390.189us ---> 38% higher throughput
integer mul: 193.532us ---> 57% higher throughput
integer lsr: 179.929us ---> 149% higher throughput

Test Plan:
Correctness:

buck2 test mode/opt //caffe2/test:test_ops

Performance:

buck2 run mode/opt //caffe2/benchmarks/operator_benchmark/fb:operator_benchmark_test

Reviewed By: mcfi

Differential Revision: D84424638
